crates/phala-allocator/src/lib.rs

```diff
@@ -0,0 +1,89 @@
+#![no_std]
+use core::alloc::{GlobalAlloc, Layout};
+use core::sync::atomic::{AtomicUsize, Ordering};
+
+pub struct StatSizeAllocator<T> {
+    inner: T,
+    current_used: AtomicUsize,
+    peak_used: AtomicUsize,
+}
+
+#[derive(Debug)]
+pub struct Stats {
+    /// The current heap usage of the allocator.
+    pub current_used: usize,
+    /// The peak heap usage of the allocator.
+    pub peak_used: usize,
+}
+
+impl<T> StatSizeAllocator<T> {
+    pub const fn new(inner: T) -> Self {
+        Self {
+            inner,
+            current_used: AtomicUsize::new(0),
+            peak_used: AtomicUsize::new(0),
+        }
+    }
+
+    pub fn stats(&self) -> Stats {
+        Stats {
+            current_used: self.current_used.load(Ordering::Relaxed),
+            peak_used: self.peak_used.load(Ordering::Relaxed),
+        }
+    }
+}
+
+impl<T: GlobalAlloc> StatSizeAllocator<T> {
+    fn add_alloced_size(&self, size: usize) {
+        let prev = self.current_used.fetch_add(size, Ordering::SeqCst);
+        let total_size = prev + size;
```

Comment:
Why adding twice? I think it's `total_size = current_used + size + size`

Reply:
> Why adding twice? I think it's `total_size = current_used + size + size`

the `fn fetch_add` returns the previous value rather than the add result.

Reply:
Ah, cool